src/util/queryNodeMatchers.ts

```diff
@@ -0,0 +1,147 @@
+import { Position, Range } from "vscode";
+import { SyntaxNode, Query, Language, QueryCapture } from "web-tree-sitter";
+import {
+  NodeMatcher,
+  NodeMatcherValue,
+  SelectionExtractor,
+  SelectionWithEditor,
+} from "../typings/Types";
+import { simpleSelectionExtractor } from "./nodeSelectors";
+
+export function defaultMatcher(
+  scopeType: string,
+  searchScopePresent: boolean,
+  scopeQuery: string,
+  selector: SelectionExtractor = simpleSelectionExtractor
+): NodeMatcher {
+  let query: Query;
+
+  return (
+    selection: SelectionWithEditor,
+    node: SyntaxNode,
+    siblings: boolean | undefined
+  ): NodeMatcherValue[] | null => {
+    if (!query) {
+      query = getQuery(node, scopeQuery);
+    }
+
+    const captures = extractCaptures(
+      selection,
+      node.tree.rootNode,
+      query,
+      scopeType
+    );
+
+    if (captures.length === 0) {
+      return null;
+    }
+    let capture: QueryCapture;
+    capture = selectCaptureByRange(captures, selection);
+    if (siblings) {
+      let captures: QueryCapture[];
+      if (searchScopePresent) {
+        throw new Error("searchScope based queries are not implemented.");
+      } else {
+        captures = findBySiblingsByParent(
+          capture.node,
+          selection,
+          query,
+          scopeType
+        );
+      }
+
+      return captures!.map((c) => {
+        return {
+          node: c.node,
+          selection: selector(selection.editor, c.node),
+        };
+      });
+    } else {
+      return [
+        {
+          node: capture!.node,
+          selection: selector(selection.editor, capture!.node),
+        },
+      ];
+    }
+  };
+}
+
+function extractCaptures(
+  selection: SelectionWithEditor,
+  root: SyntaxNode,
+  query: Query,
+  scopeType: string
+) {
+  const startPoint = generatePointFromSelection(selection, "start");
+  const endPoint = generatePointFromSelection(selection, "end");
+
+  return query.captures(root, startPoint, endPoint).filter((capture) => {
```

Comment:
It's worth having tests with various versions of overlapping selections to test this and your subsequent intersection code

- [ ] empty selection in white space before scope
- [ ] empty selection touching start of scope
- [ ] empty selection contained by scope
- [ ] empty selection touching end of scope
- [ ] empty selection in white space after scope
- [ ] nonempty selection with start and end in white space before scope
- [ ] nonempty selection with start before scope and end touching start of scope
- [ ] nonempty selection with start before scope and end contained by scope
- [ ] Etc etc
- [ ] Cases where start is in one scope and end is in another

Reply:
Heyo @pokey — for these `Error` cases, I'm going to look at units tests for them. Right now, we nullify all tests where an `Error` is thrown in `CommandRunner`.